File: src/main/java/me/oriharel/machinery/inventory/InventoryItem.java
package me.oriharel.machinery.inventory;

import me.oriharel.machinery.utilities.Callback;
import me.oriharel.machinery.utilities.NMS;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.Arrays;
import java.util.Objects;

public class InventoryItem extends ItemStack {
    protected int indexInInventory;
    protected Callback onClick = null;

    public InventoryItem(int indexInInventory, Material material, int amount, String displayName, String... lore) {
        super(material, amount);
        ItemMeta meta = NMS.getItemStackMetaReference(this);
        meta.setDisplayName(displayName);
        meta.setLore(Arrays.asList(lore));
        this.indexInInventory = indexInInventory;

    }

    public InventoryItem(int indexInInventory, ItemStack itemStack) {
        super(itemStack);
        this.indexInInventory = indexInInventory;
    }

    public InventoryItem(int indexInInventory, Material material, int amount, String displayName) {
        super(material, amount);
        ItemMeta meta = NMS.getItemStackMetaReference(this);
        meta.setDisplayName(displayName);
        this.indexInInventory = indexInInventory;
    }

    public InventoryItem runOnClick() {
        if (onClick == null) return this;
        onClick.apply();
        return this;
    }

    public InventoryItem setOnClick(Callback onClick) {
        this.onClick = onClick;
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        InventoryItem that = (InventoryItem) o;
        return indexInInventory == that.indexInInventory &&
                Objects.equals(onClick, that.onClick);
    }

    public boolean equals(ItemStack itemStack, int indexInInventory) {
        if (itemStack == null) return false;
        return itemStack.equals(this) && indexInInventory == this.indexInInventory;
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), indexInInventory, onClick);
    }
}

File: src/main/kotlin/me/oriharel/machinery/inventory/InventoryItem.kt
package me.oriharel.machinery.inventory

import me.oriharel.machinery.utilities.NMS
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import java.util.*

open class InventoryItem : ItemStack {
    var indexInInventory: Int
    private var onClick: (() -> Unit)? = null

    constructor(indexInInventory: Int, material: Material?, amount: Int, displayName: String?, vararg lore: String?) : super(material!!, amount) {
        val meta = NMS.getItemStackMetaReference(this)
        meta!!.setDisplayName(displayName)
        meta.lore = Arrays.asList(*lore)
        this.indexInInventory = indexInInventory
    }

    constructor(indexInInventory: Int, itemStack: ItemStack?) : super(itemStack!!) {
        this.indexInInventory = indexInInventory
    }

    constructor(indexInInventory: Int, material: Material?, amount: Int, displayName: String?) : super(material!!, amount) {
        val meta = NMS.getItemStackMetaReference(this)
        meta!!.setDisplayName(displayName)
        this.indexInInventory = indexInInventory
    }

    fun runOnClick(): InventoryItem {
        if (onClick == null) return this
        onClick?.invoke()
        return this
    }

    fun setOnClick(onClick: (() -> Unit)): InventoryItem {
        this.onClick = onClick
        return this
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        if (!super.equals(o)) return false
        val that = o as InventoryItem
        return indexInInventory == that.indexInInventory &&
                onClick == that.onClick
    }

    fun equals(itemStack: ItemStack?, indexInInventory: Int): Boolean {
        return if (itemStack == null) false else itemStack == this && indexInInventory == this.indexInInventory
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), indexInInventory, onClick)
    }
}